### DISCUSSION TITLE
In container, running as a jupyter notebook, but could not connect to klive server.

### DISCUSSION THREAD

**Discussion by Corafan** [gdsfactory/gdsfactory#2965] [Q&A] [answered]
Hi! 
In VS Code, I open the development container, and open a file in gdsfactory folder as a jupyter notebook, then run it. But there is an error coming up. 
![error](https://github.com/gdsfactory/gdsfactory/assets/156647994/d1b16072-2a33-457d-aefa-b6e0127976cd)
Actually, I have already installed the klive in klayout.
![klive](https://github.com/gdsfactory/gdsfactory/assets/156647994/bd61a7b6-b38d-4139-a03d-3fd622f15786)
How can I solve this problem?

**Comment by fp-matt** [comment]
What does the klive toolbar icon look like for you?

If it's otherwise running fine, it should be green like this:
![image](https://github.com/gdsfactory/gdsfactory/assets/169110167/4d88b9fb-c321-4840-bc81-a703a0628d7b)

Otherwise, I would try following the troubleshooting guide on the klive github docs and post whatever tracebacks/errors you get here.
At the very least, it might help give more information as to where things are going wrong.

https://gdsfactory.github.io/klive/troubleshooting/

**Comment by Corafan** [comment]
My icon is always green，and I tried the troubleshooting klive. But my problem is still there.
![troubleshooting guide](https://github.com/gdsfactory/gdsfactory/assets/156647994/6ee35fd7-13c4-49d4-b928-5af585fff197)

**Comment by tvt173** [comment, marked as answer]
as far as i know, klive does not currently support remote connections. 

when using a development container, gdsfactory is essentially running on a remote machine. but klayout is running on your local machine. currently i believe klive only works if both the python process and klayout are running on the same machine

**Reply by Corafan** [reply]
Ok, I see.  Thank you for your reply!

**Reply by sebastian-goeldi** [reply]
Klive would be less of a problem. It would be trivial to allow it to listen on a different port.

The problem is much more the gdsfactory container. The problem is, talking to the host and not just tunneling through (i.e. connect to an open port on the host) is a pita and doesn't work reliably afaik

**Comment by sapar82** [comment]
I believe you could forward the port from your VM to your system. I had similar issues running different scripts on Windows Subsystem for Linux. I asked O1-mini some info, hope this can help:
https://chatgpt.com/c/66fd45d7-5f58-800a-aec9-07c08188db6c

**Reply by sebastian-goeldi** [reply]
The problem is though, here you need to access something that is only running against localhost on the host (and especially it's docker, so accessing the host of docker is less than trivial and also highly depends on the docker network used plus the host OS).

So, what you need to do is access `127.0.0.1:8081` of the host of docker. `127.0.0.1` is not a properly routable IP as this should always be a loopback IP.

Now, you could have klive run on `0.0.0.0:8081` (or your global IP), but then you need a firewall which blocks properly from other sources than your dev container.

As you can imagine this is anything but easy and I'd sooner recommend doing something like kweb than this.

What is most likely easier is to have the build folder be a mounted docker volume and rely on klayout's changed gds file detection (or just plain `Ctrl + R` to reload manually)

P.S. tried to access the 20$/month friend's link, but it gave an error and couldn't find the conversation.
